node/pkg/telemetry/loki.go

```diff
@@ -142,13 +153,35 @@ func NewLokiCloudLogger(ctx context.Context, logger *zap.Logger, url string, pro
 		lokiLabels[level] = levLabels
 	}
 
+	// Create a buffered channel so the application does not block in the logger. The size of this
+	// channel should be big enough to handle a burst of log messages from the app.
+	localChan := make(chan api.Entry, 1000)
+
+	// Create a local context with a cancel function so we can signal our go routine to shutdown when the time comes.
+	localContext, localCancel := context.WithCancel(ctx)
+
+	// Kick off a go routine to read from the local buffered channel and write to the Loki unbuffered channel.
+	go func(ctx context.Context, c client.Client, localC chan api.Entry) {
+		for {
+			select {
+			case <-ctx.Done():
+				c.Stop()
+				return
+			case entry := <-localC:
+				c.Chan() <- entry // can_block: That's why we are in this separate go routine.
+			}
+		}
+	}(localContext, c, localChan)
```

Comment:
I'm wondering about the unwinding here if the cancel context is called. Similar to the Done case, what happens to the buffered messages? Do they get dropped?

Reply:
This code has changed drastically. Please take another look.